src/Confluent.Kafka/Consumer.cs

```diff
@@ -951,8 +953,8 @@ internal Consumer(ConsumerBuilder<TKey, TValue> builder)
                             },
                             IsPartitionEOF = false
                         },
-                        new Error(ErrorCode.Local_KeyDeserialization, ex.ToString())
-                    );
+                        new Error(ErrorCode.Local_KeyDeserialization),
+                        exception);
```

Comment:
The inner exception structure here allows the deserializers to throw whatever they like, whilst still providing predictability to users of the `Consume` method (they only need to test for `ConsumeException`).

It's debatable whether deserialization errors should be exposed on something that derives from `KafkaException` which is 1:1 with a librdkafka error otherwise. The main advantage is it means that users of `Consume` can just check for one very specific exception type (`ConsumeException`), otherwise it'd be `Exception` if they just want one catch block. The other advantage is the functionality of the `ConsumeException` is a perfect fit for what we need here (the `ConsumeRecord` field).

Reply:
Something I'm not 100% sure about: In the future, we could add very specific exception types that are 1:1 with librdkafka error codes. These would derive from `ConsumeException`, `ProduceException` or `KafkaException`. There is a problem here though: if a given error code can be thrown during one or more of consume, produce or outside the context of either - the existence of the produce and consume exception types in the hierarchy mean that we'd need more than on exception type per librdkafka error code. we don't want that. I haven't thought this through at all, but I'm pretty sure this is the case.

so i think we face a choice of having the features the produce and consume exception types provide (basically the consume result / delivery report info) or fine grained exception types. i think the choice is clear - we want the consume and produce exception types, and as I said, i think we're locking in the decision here that we won't ever provide finer grained exception types (at least comprehensively).

practical consequence is that to handle different librdkafka errors differently, people would need to do that with control logic inside a `catch (KafkaException)` block, not using multiple catch statements.

Reply:
I too believe detailed librdkafka 1:1 exceptions are the most user friendly.

Reply:
as discussed above, the use of `ConsumeException` and `ProduceException` may make this problematic.